runtime/server/templated_strings.go

```diff
@@ -0,0 +1,269 @@
+package server
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig/v3"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime"
+	"github.com/rilldata/rill/runtime/metricsview"
+	"github.com/rilldata/rill/runtime/metricsview/executor"
+	"github.com/rilldata/rill/runtime/parser"
+	"github.com/rilldata/rill/runtime/pkg/observability"
+	"github.com/rilldata/rill/runtime/resolvers"
+	"github.com/rilldata/rill/runtime/server/auth"
+	"go.opentelemetry.io/otel/attribute"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (s *Server) ResolveTemplatedString(ctx context.Context, req *runtimev1.ResolveTemplatedStringRequest) (*runtimev1.ResolveTemplatedStringResponse, error) {
+	s.addInstanceRequestAttributes(ctx, req.InstanceId)
+	observability.AddRequestAttributes(ctx,
+		attribute.String("args.instance_id", req.InstanceId),
+		attribute.Bool("args.use_format_tokens", req.UseFormatTokens),
+	)
+
+	claims := auth.GetClaims(ctx, req.InstanceId)
+	if !claims.Can(runtime.ReadAPI) {
+		return nil, status.Errorf(codes.FailedPrecondition, "does not have access to query data")
+	}
+
+	inst, err := s.runtime.Instance(ctx, req.InstanceId)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	templateData := parser.TemplateData{
+		User:      claims.UserAttributes,
+		Variables: inst.ResolveVariables(false),
+		State:     make(map[string]any),
+	}
+
+	funcMap := sprig.TxtFuncMap()
+	delete(funcMap, "env")
+	delete(funcMap, "expandenv")
+
+	// Register the metrics_sql custom function
+	funcMap["metrics_sql"] = func(sql string) (string, error) {
+		value, metricsViewName, fieldName, err := s.executeMetricsSQL(ctx, req.InstanceId, claims, sql, req.AdditionalWhereByMetricsView, req.AdditionalTimeRange)
+		if err != nil {
+			return "", err
+		}
+
+		// Return format token or raw value based on request
+		if req.UseFormatTokens {
+			return fmt.Sprintf(`__RILL__FORMAT__(%q, %q, %v)`, metricsViewName, fieldName, value), nil
+		}
+
+		return fmt.Sprintf("%v", value), nil
+	}
+
+	// Resolve the template
+	tmpl, err := template.New("templated_string").Funcs(funcMap).Parse(req.Data)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to parse template: %s", err.Error())
+	}
+
+	var result strings.Builder
+	err = tmpl.Execute(&result, templateData)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to execute template: %s", err.Error())
+	}
```

Comment:
Call `ResolveTemplate` instead of duplicating the template resolution logic. Since you need to inject an extra function, you can add an option for that in `parser.TemplateData` (e.g. `parser.TemplateData.ExtraFuncs`)